coderd/workspaceportshare.go

```diff
@@ -0,0 +1,137 @@
+// @Success 200
+// @Router /workspaces/{workspace}/port-share [post]
+func (api *API) postWorkspaceAgentPortShare(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	workspace := httpmw.WorkspaceParam(r)
+	portSharer := *api.PortSharer.Load()
+	var req codersdk.UpdateWorkspaceAgentPortShareRequest
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	if req.ShareLevel < 0 || req.ShareLevel > 2 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Port sharing level not allowed. Must be between 0 and 2.",
+		})
+		return
+	}
+
+	if portSharer.CanRestrictSharing() {
+		template, err := api.Database.GetTemplateByID(ctx, workspace.TemplateID)
+		if err != nil {
+			httpapi.InternalServerError(rw, err)
+			return
+		}
+
+		if req.ShareLevel > codersdk.WorkspaceAgentPortShareLevel(template.MaxPortSharingLevel) {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: fmt.Sprintf("Port sharing level not allowed. Must not be greater than %d.", template.MaxPortSharingLevel),
+			})
+			return
+		}
+	}
+
+	agents, err := api.Database.GetWorkspaceAgentsInLatestBuildByWorkspaceID(ctx, workspace.ID)
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+
+	found := false
+	for _, agent := range agents {
+		if agent.Name == req.AgentName {
+			found = true
+			break
+		}
+	}
+	if !found {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Agent not found.",
+		})
+		return
+	}
+
+	psl, err := api.Database.GetWorkspaceAgentPortShare(ctx, database.GetWorkspaceAgentPortShareParams{
+		WorkspaceID: workspace.ID,
+		AgentName:   req.AgentName,
+		Port:        req.Port,
+	})
+	if err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			httpapi.InternalServerError(rw, err)
+			return
+		}
+
+		if req.ShareLevel == codersdk.WorkspaceAgentPortShareLevelOwner {
+			// If the port is not shared, and the user is trying to set it to owner,
+			// we don't need to do anything.
+			rw.WriteHeader(http.StatusOK)
+			return
+		}
+
+		_, err = api.Database.InsertWorkspaceAgentPortShare(ctx, database.InsertWorkspaceAgentPortShareParams{
+			WorkspaceID: workspace.ID,
+			AgentName:   req.AgentName,
+			Port:        req.Port,
+			ShareLevel:  int32(req.ShareLevel),
+		})
+		if err != nil {
+			httpapi.InternalServerError(rw, err)
+			return
+		}
+
+		rw.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if codersdk.WorkspaceAgentPortShareLevel(psl.ShareLevel) == codersdk.WorkspaceAgentPortShareLevelOwner {
+		// If the port is shared, and the user is trying to set it to owner,
+		// we need to remove the existing share record.
+		err = api.Database.DeleteWorkspaceAgentPortShare(ctx, database.DeleteWorkspaceAgentPortShareParams{
+			WorkspaceID: workspace.ID,
+			AgentName:   req.AgentName,
+			Port:        req.Port,
+		})
+		if err != nil {
+			httpapi.InternalServerError(rw, err)
+			return
+		}
+
+		rw.WriteHeader(http.StatusOK)
+		return
+	}
```

Comment:
IMO the UI should not show "owner" as a sharing option, and should just have "authenticated" and "public" as options, with a delete button. The API shouldn't accept "owner" either

Reply:
If you use the existing enum in the DB, make sure you add a constraint that prevents the value from being owner.